integrations/terraform/provider/credentials.go

```diff
@@ -0,0 +1,421 @@
+				brokenCredentialErrorSummary(source.Name(), reason, err),
+			)
+			return nil, diags
+		}
+
+		// Smoke test to see if the credential is valid
+		// This catches all the "file not found" issues and other broken credentials
+		// so we can turn them into a hard failure.
+		_, err = creds.TLSConfig()
+		if err != nil {
+			logFields["error"] = err.Error()
+			tflog.Error(ctx, "failed to get a TLSConfig from the credential", logFields)
+			_, reason := source.IsActive(providerCfg)
+			diags.AddError(
+				fmt.Sprintf("Invalid Teleport credentials %s", source.Name()),
+				brokenCredentialErrorSummary(source.Name(), reason, err),
+			)
+
+			return nil, diags
+		}
+
+		now := time.Now()
+		if expiry, ok := creds.Expiry(); ok && !expiry.IsZero() && expiry.Before(now) {
+			diags.AddWarning(
+				fmt.Sprintf("Teleport credentials %s are expired", source.Name()),
+				fmt.Sprintf(`The credentials %s are expired. Expiration is %q while current time is %q). You might need to refresh them. The provider will not attempt to use those credentials.`,
+					source.Name(), expiry.Local(), now.Local()),
+			)
+			continue
+		}
+
+		clientCfg.Credentials = []client.Credentials{creds}
+		// In case of connection failure, this takes 30 seconds to return, which is very, very long.
+		clt, err := client.New(ctx, clientCfg)
+		if err != nil {
+			logFields["error"] = err.Error()
+			tflog.Error(ctx, "failed to connect with the credential", logFields)
+			diags.AddWarning(
+				fmt.Sprintf("Failed to connect with credentials %s", source.Name()),
+				fmt.Sprintf("The client built from the credentials %s failed to connect to %q with the error: %s.",
+					source.Name(), clientCfg.Addrs[0], err,
+				))
+			continue
+		}
+		// A client was successfully built
+		return clt, diags
+	}
+	// No client was built
+	diags.AddError("Impossible to build Teleport client", s.failedToBuildClientErrorSummary(clientCfg.Addrs[0]))
+	return nil, diags
+}
+
+// failedToBuildClientErrorSummary builds a user-friendly message explaining we failed to build a functional Teleport
+// client and listing every connection method we tried.
+func (s CredentialSources) failedToBuildClientErrorSummary(addr string) string {
+	sb := strings.Builder{}
+	sb.WriteString("Every credential source provided has failed. The Terraform provider cannot connect to the Teleport cluster '")
+	sb.WriteString(addr)
+	sb.WriteString("'.\n\n")
+	sb.WriteString("We tried building a client:\n")
+	for _, source := range s {
+		sb.WriteString(" - ")
+		sb.WriteString(source.Name())
+		sb.WriteRune('\n')
+	}
+	sb.WriteRune('\n')
+	sb.WriteString("You can find more information about each credential source specific errors in the Terraform warnings above this error.")
+	return sb.String()
+}
+
+// brokenCredentialErrorSummary returns a user-friendly message expalining why we failed to
+func brokenCredentialErrorSummary(name, activeReason string, err error) string {
```

Comment:
What do you think about using text/template here? I think it would improve the legibility.